engine/src/main/java/io/camunda/zeebe/streamprocessor/DirectProcessingResultBuilder.java

```diff
@@ -54,6 +68,21 @@ public ProcessingResultBuilder appendRecord(
       final RejectionType rejectionType,
       final String rejectionReason,
       final RecordValue value) {
+
+    final ValueType valueType = typeRegistry.get(value.getClass());
+    if (valueType == null) {
+      // usually happens when the record is not registered at the TypedStreamEnvironment
+      throw new IllegalStateException("Missing value type mapping for record: " + value.getClass());
+    }
```

Comment:
This is right now similar handled in the legacy writer https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/streamprocessor/LegacyTypedStreamWriterImpl.java#L40-L45